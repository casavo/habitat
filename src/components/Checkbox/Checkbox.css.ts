import { globalStyle, style } from "@vanilla-extract/css";
import { vars } from "../../utils/theme.css";
import { recipe } from "@vanilla-extract/recipes";

const gap = vars.space[100];
const size = vars.space[200];
const black = vars.colors.neutral[100];
const grey = "#B3B3B2";
const red = vars.colors.alertcolors.error["100"];

export const wrapper = style({
  alignItems: "center",
  display: "flex",
  flexWrap: "wrap",
  gap: `0 ${gap}`,
  transition: "all 200ms",
});

export const checkbox = recipe({
  base: {
    alignItems: "center",
    backgroundColor: vars.colors.neutral[0],
    border: `1px solid ${grey}`,
    borderRadius: 4,
    display: "flex",
    height: size,
    justifyContent: " center",
    transition: "all 200ms",
    width: size,
  },
  compoundVariants: [
    {
      style: {
        backgroundColor: red,
        borderColor: red,
      },
      variants: {
        checked: true,
        error: true,
      },
    },
  ],
  variants: {
    checked: {
      true: {
        backgroundColor: black,
        borderColor: black,
      },
    },
    disabled: {
      true: {
        backgroundColor: "#E8E8E8",
        borderColor: grey,
      },
    },
    error: {
      true: {
        borderColor: red,
      },
    },
  },
});

globalStyle(`div[data-id="check-icon"] > svg`, {
  fill: "none",
});

globalStyle(`${wrapper}[data-disabled] > p`, {
  color: grey,
});

export const messageStyle = style({
  color: red,
  flexBasis: "100%",
  margin: 0,
  marginLeft: `calc(${size} + ${gap})`,
});
