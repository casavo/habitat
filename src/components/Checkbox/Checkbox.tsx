export const Checkbox = () => <input type="checkbox" />;
