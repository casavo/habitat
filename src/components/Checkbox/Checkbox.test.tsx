import { fireEvent, render, screen } from "@testing-library/react";
import { expect, describe, vi } from "vitest";

import { Checkbox } from "./Checkbox.tsx";
import {ComponentProps} from "react";


type onChangeType = NonNullable<ComponentProps<typeof Checkbox>['onChange']>
describe("Checkbox component", () => {
  it("renders", () => {
    const { container } = render(<Checkbox />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders as selected", () => {
    const { container } = render(<Checkbox checked />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders as disabled", () => {
    const { container } = render(<Checkbox disabled />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("calls onChange", async () => {
    const onChange = vi.fn<Parameters<onChangeType>,void>();
    render(<Checkbox onChange={onChange} />);
    const input = screen.getByRole("checkbox");

    fireEvent.click(input);
    expect(onChange).toHaveBeenCalledWith(true);
  });

  it("calls onChange when is already checked", async () => {

    const onChange = vi.fn<Parameters<onChangeType>,void>();
    render(<Checkbox checked={true} onChange={onChange} />);
    const input = screen.getByRole("checkbox");

    fireEvent.click(input);
    expect(onChange).toHaveBeenCalledWith(false);
  });
});
