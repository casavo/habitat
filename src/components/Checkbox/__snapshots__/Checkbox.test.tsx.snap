// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checkbox component > renders 1`] = `
<label
  class="Checkbox_wrapper__1g2c7v70"
  data-rac=""
>
  <span
    style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
  >
    <input
      type="checkbox"
    />
  </span>
  <div
    class="Checkbox_checkbox__1g2c7v71"
    data-id="check-icon"
  />
  <p
    class="Typography__3igy4nb Typography_base__3igy4na Typography_BodyStyle_noMargin_true__3igy4nc Typography_BodyStyle_size_m__3igy4ne"
  />
</label>
`;

exports[`Checkbox component > renders as disabled 1`] = `
<label
  class="Checkbox_wrapper__1g2c7v70"
  data-disabled="true"
  data-rac=""
>
  <span
    style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
  >
    <input
      disabled=""
      type="checkbox"
    />
  </span>
  <div
    class="Checkbox_checkbox__1g2c7v71 Checkbox_checkbox_disabled_true__1g2c7v73"
    data-id="check-icon"
  />
  <p
    class="Typography__3igy4nb Typography_base__3igy4na Typography_BodyStyle_noMargin_true__3igy4nc Typography_BodyStyle_size_m__3igy4ne"
  />
</label>
`;

exports[`Checkbox component > renders as selected 1`] = `
<label
  class="Checkbox_wrapper__1g2c7v70"
  data-rac=""
>
  <span
    style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
  >
    <input
      type="checkbox"
    />
  </span>
  <div
    class="Checkbox_checkbox__1g2c7v71"
    data-id="check-icon"
  />
  <p
    class="Typography__3igy4nb Typography_base__3igy4na Typography_BodyStyle_noMargin_true__3igy4nc Typography_BodyStyle_size_m__3igy4ne"
  />
</label>
`;
